### PR TITLE
util/types: properly handle truncate error for string to float or int.

### PR DESCRIPTION
**Pull request by coocood**
When converting from string to float or int, use StatementContext to handle truncate error.
Also directly convert a string to Int instead converting to float first, then convert float to int.

### CODE REVIEW

File: util/types/convert.go [hunk 1]
```diff
@@ -138,78 +147,151 @@ func isCastType(tp byte) bool {
...
+		case '.':
+			dotIdx = i
+		case 'e', 'E':
+			eIdx = i
 		}
-		r = r*10 + int64(str[i]-'0')
 	}
-	if negative {
-		r = -r
+	if dotIdx == -1 && eIdx == -1 {
```
Comment (tiancaiamao) [inline util/types/convert.go]: what if the input string contains multi 'e' or '.' ?

Comment (coocood) [inline util/types/convert.go]: Then it is not valid float string.
This function only handle valid float string.

File: util/types/convert.go [hunk 2]
```diff
@@ -138,78 +147,151 @@ func isCastType(tp byte) bool {
...
+	var validInt string
+	if intCnt <= len(digits) {
+		validInt = string(digits[:intCnt])
+	} else {
+		validInt = string(digits) + strings.Repeat("0", intCnt-len(digits))
+	}
+	return validInt, nil
 }
 
 // StrToFloat converts a string to a float64 in best effort.
```
Comment (tiancaiamao) [inline util/types/convert.go]: I think it should be

> // StrToFloat converts a string to a float64 **at the best effort**.

@queenypingcap

File: util/types/convert.go [hunk 3]
```diff
@@ -137,79 +134,152 @@ func isCastType(tp byte) bool {
...
+	str = strings.TrimSpace(str)
+	validPrefix, err := getValidIntPrefix(sc, str)
+	if validPrefix[0] == '+' {
+		validPrefix = validPrefix[1:]
+	}
+	uVal, err1 := strconv.ParseUint(validPrefix, 10, 64)
+	if err1 != nil {
+		return uVal, errors.Trace(ErrOverflow)
+	}
+	return uVal, err
```
Comment (shenli) [inline util/types/convert.go]: errors.Trace(err)

File: util/types/convert.go [hunk 4]
```diff
@@ -137,79 +146,152 @@ func isCastType(tp byte) bool {
...
-	} else if str[i] == '+' {
-		i++
-	}
-	r := int64(0)
-	for ; i < len(str); i++ {
-		if !unicode.IsDigit(rune(str[i])) {
-			break
+	validPrefix, err := getValidIntPrefix(sc, str)
+	iVal, err1 := strconv.ParseInt(validPrefix, 10, 64)
+	if err1 != nil {
```
Comment (coocood) [inline util/types/convert.go]: Valid prefix never return truncate error.

File: util/types/convert.go [hunk 5]
```diff
@@ -137,79 +146,152 @@ func isCastType(tp byte) bool {
...
+// StrToUint converts a string to an unsigned interger at the best-effortt.
+func StrToUint(sc *variable.StatementContext, str string) (uint64, error) {
+	str = strings.TrimSpace(str)
+	validPrefix, err := getValidIntPrefix(sc, str)
+	if validPrefix[0] == '+' {
+		validPrefix = validPrefix[1:]
+	}
+	uVal, err1 := strconv.ParseUint(validPrefix, 10, 64)
+	if err1 != nil {
+		return uVal, errors.Trace(ErrOverflow)
```
Comment (coocood) [inline util/types/convert.go]: dito

File: util/types/convert.go [hunk 6]
```diff
@@ -137,79 +146,152 @@ func isCastType(tp byte) bool {
...
+	}
+	return uVal, err
+}
+
+// getValidIntPrefix gets prefix of the string which can be successfully parsed as int.
+func getValidIntPrefix(sc *variable.StatementContext, str string) (string, error) {
+	floatPrefix, err := getValidFloatPrefix(sc, str)
```
Comment (coocood) [inline util/types/convert.go]: Because valid floatPrefix like `1.1e10` can not be parsed by `strconv.ParseInt`.

**Comment by shenli** [comment]
LGTM

File: util/types/convert_test.go [hunk 7]
```diff
@@ -532,6 +570,10 @@ func (s *testTypeConvertSuite) TestConvert(c *C) {
 	signedAccept(c, mysql.TypeFloat, float64(123), "123")
 	signedAccept(c, mysql.TypeDouble, " -23.54", "-23.54")
 	signedDeny(c, mysql.TypeDouble, "-23.54a", "-23.54")
+	signedDeny(c, mysql.TypeDouble, "-23.54e2e", "-2354")
+	signedDeny(c, mysql.TypeDouble, "+.e", "0")
+	signedAccept(c, mysql.TypeDouble, "1e+1", "10")
+	signedAccept(c, mysql.TypeDouble, "1e+1", "10")
```
Comment (zimulala) [inline util/types/convert_test.go]: Duplicated line.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**